crates/node-bindings/src/core/js_requests/request_options.rs

```diff
@@ -0,0 +1,43 @@
+//! Provides helpers to cast from a `JsObject` options object into a few common
+//! options
+//!
+//! This corresponds to the `RequestOptions` javascript type.
+//!
+//! The options read are `options.inputFS` and `options.projectRoot`.
+//!
+//! This is either a no-copy (for inputFS) or a copy on read operation
+//! (for projectRoot).
+use std::rc::Rc;
+
+use napi::Env;
+use napi::JsObject;
+use napi::JsString;
+use parcel_filesystem::js_delegate_file_system::JSDelegateFileSystem;
+
+pub fn project_root_from_options(options: &JsObject) -> napi::Result<String> {
+  let Some(project_root): Option<JsString> = options.get("projectRoot")? else {
+    return Err(napi::Error::from_reason(
+      "[napi] Missing required projectRoot options field",
+    ));
+  };
+  let project_root = project_root.into_utf8()?;
+  let project_root = project_root.as_str()?;
+  Ok(project_root.to_string())
+}
+
+pub fn input_fs_from_options(
+  env: Rc<Env>,
+  options: &JsObject,
+) -> napi::Result<JSDelegateFileSystem> {
+  let Some(input_fs) = options
+    .get("inputFS")?
+    .map(|input_fs| JSDelegateFileSystem::new(env, input_fs))
+  else {
+    // We need to make the `FileSystem` trait object-safe, so we can use dynamic
```

Comment:
Note here;

At the moment `FileSystem` has a few generics which make it non object safe.

We want to simplify it so that we can make it object-safe. Making it object safe will allow us to pass around a `Box<dyn FileSystem>`, which would do dynamic dispatch.

This is good because we can have some fallbacks. For example, if JavaScript doesn't give us a filesystem, it is a sensible option to use the `OsFileSystem` implementation, since that's eventually what we will want, rather than JavaScript providing us a real `FileSystem`.